### PR TITLE
integration-tests: add integration test for local-gadget

### LINKED ISSUE #738: [RFE] local-gadget: Create integration tests
## Current situation

There are no integration-tests for the `local-gadget`. Once we agree on how to implement the single call run https://github.com/kinvolk/inspektor-gadget/pull/730, we should start implementing the tests for them.  

As said in https://github.com/kinvolk/inspektor-gadget/issues/719, these tests should test the JSON format and the transformation from JSON to columns will be done by unit-testing the GadgetParsers' method, e.g. for [the socket gadget](https://github.com/kinvolk/inspektor-gadget/blob/65a398e8ef72887c9b36550de64e7b00cfa8451f/cmd/kubectl-gadget/snapshot/socket.go#L33-L38).

### PR DESCRIPTION
**Pull request by mqasimsarfraz**
Add support for running integration test for local-gadget against multiple container runtimes.

Fixes #738 

### Idea
The idea is based on following changes:
- Make `command.go` API public to share between different test environments.
- Use minikube to spin-up different container runtime environments.
- Hook the test in CI to run on each commit.
- Add couple of test to ensure the setup works properly. 

### Testing Done
Did a manual run of integration test against an AKS cluster.

### CODE REVIEW

File: integration/local-gadget/helpers_test.go
Comment (mqasimsarfraz) [inline integration/local-gadget/helpers_test.go:36]: I haven't used it anywhere yet but since docker also supports CRI so having this generic to way to create/remove test container should work fine for us.

File: integration/local-gadget/helpers_test.go
Comment (mqasimsarfraz) [inline integration/local-gadget/helpers_test.go:18]: It would be nice to use [command](https://github.com/kinvolk/inspektor-gadget/blob/main/integration/command.go#L36) here. But I wanted to avoid adding multi module dependency like [this](https://github.com/kinvolk/inspektor-gadget/pull/945#discussion_r972908078). Also #945 will be super helpful for JSON validation.

**Review by mqasimsarfraz — COMMENTED** [review]
@blanquicet I looked into the initial idea regrading running test against multiple runtimes. Would be great to get feedback if you think the change is heading in right direction?

**Review by blanquicet — COMMENTED** [review]
It is just great @mqasimsarfraz! I really like how it is already looking 💯 

The setup is very complete, clear and easy to understand. The part I like more is that it creates a solid base for the full feature. I think it is great you already thought about creating a "framework" generic enough that can support all the runtimes. Awesome! 🎉 

I think we definitely need to tackle https://github.com/kinvolk/inspektor-gadget/issues/944 first, and then we can create a set of helper functions to verify the JSON output, see my comment below.

File: integration/local-gadget/Makefile
Comment (blanquicet) [inline integration/local-gadget/Makefile:70]: I propose adding `-test.v` in this initial phase, we can remove it once the tests are more stable:

```suggestion
	$(CONTAINER_CLI) exec $${MINIKUBE_PROFILE} local-gadget-integration.test -integration -test.v -container-runtime $*
```

File: integration/local-gadget/integration_test.go
Comment (blanquicet) [inline integration/local-gadget/integration_test.go:74]: I propose adding another test here using the `--containername` flag and then using a helper function that verifies there is only one result, and it matches the expected one. While the existing test (without the filtering) expects multiple results but only one must match the expected one.

File: integration/local-gadget/integration_test.go
Comment (blanquicet) [inline integration/local-gadget/integration_test.go:81]: Could you please start using the new notation we want to adopt: https://github.com/kinvolk/inspektor-gadget/issues/881#issuecomment-1231424009?

File: integration/local-gadget/Makefile
Comment (blanquicet) [inline integration/local-gadget/Makefile:67]: I think stderr shouldn't be sent to redirected, otherwise, it will just fail, and we will not know why.

File: integration/local-gadget/helpers_test.go
Comment (mqasimsarfraz) [reply integration/local-gadget/helpers_test.go:48]: Yeah, makes much more sense and also the workflow for both the integration tests will be similar. I will switch to using `busyboxPodCommand` after #944 is merged, will make everything simpler.

File: integration/local-gadget/integration_test.go
Comment (mqasimsarfraz) [reply integration/local-gadget/integration_test.go:27]: hehe. Side effect of ctrl-c + ctrl-v :smile: But my initial idea was to do some validation like the docker version since we need specific one for cri-o

**Review by mqasimsarfraz — COMMENTED** [review]
@blanquicet thanks for an awesome review. I will rework the PR to address the suggestions!

File: .github/workflows/inspektor-gadget.yml
Comment (blanquicet) [reply .github/workflows/inspektor-gadget.yml:488]: Yeah, I like your solution. It is pretty simple and doesn't have any external depency. I was just wondering how were they managing this issue.

File: integration/local-gadget/Makefile
Comment (blanquicet) [reply integration/local-gadget/Makefile:1]: Yeah, probably removed.

File: integration/local-gadget/integration_test.go
Comment (blanquicet) [reply integration/local-gadget/integration_test.go:89]: For instance, once https://github.com/kinvolk/inspektor-gadget/pull/973 is merged, we can verify here that containers created with `kubectl-gadget` were enriched with the correct values: namespace, pod and poduid. Such verification should be omitted if containers were created via `docker`, `ctr` or ... I don't know how to create standalone CRI-O containers 😅 .

Comment (mqasimsarfraz) [reply integration/local-gadget/integration_test.go:89]: Nice idea. As of now I don't think it would be possible to create non CRI complaint container uing CRI-O. I used runtime service or `crictl` to create such containers and that obviously have the concept of pod/sandbox. It would be so weird if it does support something like that! :smile:

Comment (blanquicet) [reply integration/local-gadget/integration_test.go:89]: Yeah, I also think it is not possible. Maybe it is not even possible by definition 😅.

File: integration/local-gadget/integration_test.go
Comment (blanquicet) [inline integration/local-gadget/integration_test.go:203]:
```suggestion
					return fmt.Errorf("expect all containers to match, expected=%+v, got=%+v", expectedContainer, gotContainer)
```

File: integration/local-gadget/integration_test.go
Comment (blanquicet) [inline integration/local-gadget/integration_test.go:14]: TIL: https://go.dev/ref/spec#Import_declarations

**Review by blanquicet — COMMENTED** [review]
I like the result. It is simple and reuses most of the code already created for the inspektor-gadget tests. As I already said, I think it defines a solid base to start adding more complex tests.

Thanks for also updating the documentation. It is clear and really helps.

File: integration/local-gadget/integration_test.go
Comment (blanquicet) [inline integration/local-gadget/integration_test.go:247]: Why can `Interface` not be verified?

File: integration/local-gadget/integration_test.go
Comment (blanquicet) [inline integration/local-gadget/integration_test.go:242]: Once we have the Kubernetes metadata available, I think we should start considering moving the shared code between inspektor-gadget and local-gadget to a common place. For instance, this test is identical except for the `Cmd` that calls `kubectl-gadget` or `local-gadget`.

Also because I expect to add more tests like this for each gadget that will be identical again. For example, to verify that the flag `--pid`, `--ports` and `--ignore-errors` are working properly.

cc @mauriciovasquezbernal

File: integration/local-gadget/integration_test.go
Comment (blanquicet) [inline integration/local-gadget/integration_test.go:250]: The only filtering available is `--containername`, and given that it cannot be used for docker yet, we are capturing the events of all the containers. Given that, I suspect `ExpectAllToMatch` is working in this case because the bind events are events that are not triggered very frequently. To be sure the tests will succeed, I think we should use `ExpectEntriesToMatch` until we are able to filter the output. WDYT?

File: integration/local-gadget/integration_test.go
Comment (blanquicet) [reply integration/local-gadget/integration_test.go:247]: Yeah, there was a fix to change the default from "0" to "". Probably that's the issue you were hitting.

File: integration/local-gadget/integration_test.go
Comment (blanquicet) [reply integration/local-gadget/integration_test.go:250]: Mmmm you can mention https://github.com/kinvolk/inspektor-gadget/issues/644.

File: integration/local-gadget/integration_test.go
Comment (mqasimsarfraz) [reply integration/local-gadget/integration_test.go:99]: Sounds good. I want to get out of this rebase loop I am in right now :smile:

File: integration/local-gadget/Makefile
Comment (mqasimsarfraz) [reply integration/local-gadget/Makefile:21]: Definitely, makes sense to do that. Sounds good, I can handle this a follow up PR.

**Comment by mqasimsarfraz** [comment]
@blanquicet thanks for all the reviews. I have updated the PR to handle all the feedback!

File: integration/local-gadget/Makefile
Comment (blanquicet) [inline integration/local-gadget/Makefile:67]: Could you please add here a comment with an example of how `INTEGRATION_TESTS_PARAMS` can be used like you did for the inpektor-gadget tests?

```bash
# INTEGRATION_TESTS_PARAMS="-test.run TestListContainers" CONTAINER_RUNTIME=containerd make -C integration/local-gadget test
```

File: integration/local-gadget/integration_test.go
Comment (blanquicet) [inline integration/local-gadget/integration_test.go:184]: This logic will make pass the test if the list of the returned containers is empty.

**Review by blanquicet — COMMENTED** [review]
IMO, after fixing the logic in `TestListSingleContainer`, it is ready to be merged.

What do other folks think about it?

File: integration/local-gadget/integration_test.go
Comment (blanquicet) [inline integration/local-gadget/integration_test.go:179]: nit: `output` is already being printed by `Command.Run`.

**Review by blanquicet — APPROVED** [review]
LGTM! I really like the work you did here @mqasimsarfraz 💯 !!

Let's give some time before merging it to see if other folks have comments.

File: integration/local-gadget/Makefile
Comment (mqasimsarfraz) [reply integration/local-gadget/Makefile]: I looked up `minikube exec` and stopped there :smile:  I switched it to `minikube ssh` now, I like it that we are doing everything with minikube now. Thanks for the pointer.

**Comment by mqasimsarfraz** [comment]
@eiffel-fl Thanks for the review and the feedback. I will keep a note of this for future PRs and will split into multiple commits. Sorry for the pain :laughing: I have updated the PR to address some of the comments you had!

File: integration/local-gadget/Makefile
Comment (blanquicet) [reply integration/local-gadget/Makefile]: Agree! Great suggestion Francis!

**Comment by mqasimsarfraz** [comment]
Sounds good. I will go ahead and merge this by EOD if no one has objections. I plan to create a follow up PR for suggested improvements!

File: integration/local-gadget/integration_test.go
Comment (mauriciovasquezbernal) [inline integration/local-gadget/integration_test.go:49]: Any reason to use -1? We're using 1 in all places.

File: integration/local-gadget/integration_test.go
Comment (mauriciovasquezbernal) [inline integration/local-gadget/integration_test.go:54]: nit: You can move this to L36 and avoid the `*containerRuntime != ""` check. Something like:

```go 
	if *containerRuntime == "" {
		fmt.Println("Error: '-container-runtime' must be specified")
		os.Exit(1)
	}

	found := false
	for _, val := range supportedContainerRuntimes {
		if *containerRuntime == val {
			found = true
			break
		}
	}

	if !found {
		fmt.Fprintf(os.Stderr, "Error: invalid argument '-container-runtime': %q. Valid values: %s\n",
			*containerRuntime, strings.Join(supportedContainerRuntimes, ", "))
		os.Exit(-1)
	}
```

File: integration/local-gadget/integration_test.go
Comment (mauriciovasquezbernal) [reply integration/local-gadget/integration_test.go:100]: +1 on implementing a helper. We'll get crazy with the naming of the two different cases but let's see.

File: integration/local-gadget/integration_test.go
Comment (mauriciovasquezbernal) [inline integration/local-gadget/integration_test.go:108]:
```suggestion
		BusyboxPodCommand(ns, "sleep inf"),
```

File: integration/local-gadget/integration_test.go
Comment (mauriciovasquezbernal) [inline integration/local-gadget/integration_test.go:120]: nit: is `po` and easy to understand variable name?

File: integration/local-gadget/integration_test.go
Comment (mauriciovasquezbernal) [inline integration/local-gadget/integration_test.go:75]: I suppose this will be also handled when https://github.com/kinvolk/inspektor-gadget/issues/737 is fixed.

File: integration/local-gadget/integration_test.go
Comment (mauriciovasquezbernal) [reply integration/local-gadget/integration_test.go:231]: Yes, I agree. In this case we're testing Kubernetes containers I expected both tests to be almost identical.

File: integration/local-gadget/integration_test.go
Comment (mauriciovasquezbernal) [inline integration/local-gadget/integration_test.go:212]: Why `TraceTCPBind` instead of only `TraceBind`?

File: integration/local-gadget/integration_test.go
Comment (mauriciovasquezbernal) [inline integration/local-gadget/integration_test.go:1]: nit: license header missing.

File: integration/local-gadget/integration_test.go
Comment (mauriciovasquezbernal) [inline integration/local-gadget/integration_test.go:2]: I think splitting this file into a per gadget file will help us (https://github.com/kinvolk/inspektor-gadget/issues/974). What about splitting the file now before merging this PR?

File: integration/local-gadget/integration_test.go
Comment (mqasimsarfraz) [reply integration/local-gadget/integration_test.go]: I saw that we are using `ns` so I thought using short form from `kubectl` for pods i.e `po` would be okay. What do you say?

Comment (mauriciovasquezbernal) [reply integration/local-gadget/integration_test.go]: I didn't know `po` was common. Let's keep po, that should be fine.

**Comment by mqasimsarfraz** [comment]
@mauriciovasquezbernal thanks for the review. I have adjusted the nits. Good point regrading introducing `k8s` directory, Currently  I see we can have `docker` or `k8s` based test but wanted to have deeper look into it. Would it be okay if handle this in follow up PR ?

**Review by mauriciovasquezbernal — APPROVED** [review]

**Comment by mauriciovasquezbernal** [comment]
@mqasimsarfraz , yes, that's fine to handle in a different PR.

**Pull request merged**
